src/fsharp/lib.fs

```diff
@@ -178,14 +178,11 @@ module ListAssoc =
 //------------------------------------------------------------------------
 
 module ListSet = 
-    (* NOTE: O(n)! *)
-    let rec contains f x l = 
-        match l with 
-        | [] -> false
-        | x'::t -> f x x' || contains f x t
+    let inline contains f x l = List.exists (f x) l
```

Comment:
The real question is: can we use a different data structure!?

It needs to be a set that allows to pass in own equality semantics.